### DISCUSSION TITLE
Expose non-public schema to supabase-js client

### DISCUSSION THREAD

**Discussion by giannico** [supabase/supabase#1222] [Questions] [answered]
Is it possible to expose a custom database schema directly to the `@supabase/supabase-js` client? Can I extend the list of exposed schemas beyond `public` and `storage`?

I see the client constructor takes a schema option that defines the option as:

**options.schema** - You can switch in between schemas. The schema needs to be on the list of exposed schemas inside Supabase.

I believe I have my schema grants in order, but, when I try to make a request with the client I'm failing content negotiation with a 406 and I get the response:

```{"message":"The schema must be one of the following: public, storage"}```

**Comment by steve-chavez** [comment, marked as answer]
Not right now, but it will possible soon(ish). The supabase UI will add a `db-schemas` setting on `/settings/api`, there you'll be able to add more schemas besides `public` and `storage`.

**Reply by prescience-data** [reply]
@steve-chavez Did this end up getting implemented?

**Reply by w-gitops** [reply]
bump as well. starting a project, don't want to depend on public schema

**Reply by jeyjeycodes** [reply]
Bump as well! We DO NOT want to use public schema. I may have created a duplicate discussion point. https://github.com/supabase/supabase/discussions/3204

**Reply by MildTomato** [reply]
we are working on this now, will update this when its live.

**Reply by besteman** [reply]
@MildTomato 
Any updates on the potential live date for this?

**Reply by bnjmnt4n** [reply]
Just as an update: this has landed on the Supabase platform: https://supabase.io/docs/reference/javascript/initializing#api-schemas.

**Reply by rbrueckner** [reply]
@MildTomato: any ETA on this topic? On Sep 15, 2021 you mentioned that you are working on this, but I still don't see this working.

**Reply by TomasHubelbauer** [reply]
FWIW I get `The schema must be one of the following: public, storage, graphql_public` even when my Supabase client is authorized (so it is not because it would be using only the public anon key - it has correct auth). So while the JS SDK now has the option to configure the DB schema, it looks like it is locked to those three fixed values only?

**Reply by GaryAustin1** [reply]
@TomasHubelbauer Did you set the schema in the API settings?
![image](https://github.com/supabase/supabase/assets/54564956/85c8d001-801e-48b4-940a-e58fccb766c1)

**Reply by TomasHubelbauer** [reply]
Thanks for the tip! Does this mean a Postgres function in this schema will be callable with the public anon key or is it possible to enforce a function must be called via the `rpc` method only when authorized? (I apologize for being possible off-topic from the original thread with this.)

**Reply by GaryAustin1** [reply]
If you have follow up please start a new discussion or try Supabase Discord.  Not fair to other users on this thread to get notifications if they have it on.
It does IF you granted that role access to the schema.  You can control all of the roles that can access with Postgres grants.  By default it should only allow access from Postgres role so none of the API roles can use it.  You have to add them.  https://supabase.com/docs/guides/api/using-custom-schemas

**Comment by glomyst** [comment]
Its a very urgently needed feature. This stops us from using Supabase-Js even with a self hosted Supabase instance.
Unnecessary limitation from JS side.

**Comment by ARMATAV** [comment]
Exposes, but does not add it to the `api` documentation list as it should.

**Comment by johndpope** [comment]
note - you should be able to bypass it using vanilla postgres library

import pg from 'pg';

// node-pg returns numerics as strings by default. since we don't expect to
// have large currency values, we'll parse them as floats instead.
pg.types.setTypeParser(1700, (val) => parseFloat(val));

export const db = new pg.Pool({
  database: 'postgres',
  // host: 'db.1234.supabase.co',
  host: '0.0.0.0',
  // port: 6543,
  port: 5432,
  user: 'postgres',
  // password: 'm[`(B56\\V~SQ$/<)',
  password: 'JzSL#g89@f7wzf!',
  max: 10,
  min: 2,
  idleTimeoutMillis: 1000, // close idle clients after 1 second
  connectionTimeoutMillis: 1000, // return an error after 1 second if connection could not be established
});

**Comment by ItayShmuel** [comment]
You can expose the custom schema using the config.toml file API configuration section.
<img width="801" alt="image" src="https://github.com/user-attachments/assets/8a6c1800-17fb-421c-835a-e2d288236f0e" />

**Comment by Z0rdak** [comment]
I encountered this issue as well today with the self-hosted version (using docker)

I added my schema to the .env file:
`PGRST_DB_SCHEMAS=public,storage,graphql_public,my_custom_schema`

Then `docker compose down` followed by `docker compose up -d` and it worked.
